SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialSpanIdStore.cpp

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/Connection/SpatialSpanIdStore.h"
+
+#include "SpatialConstants.h"
+#include <WorkerSDK/improbable/c_trace.h>
+
+DEFINE_LOG_CATEGORY(LogSpatialSpanIdStore);
+
+using namespace SpatialGDK;
+using namespace worker::c;
+
+SpatialSpanIdStore::SpatialSpanIdStore()
+{
+	UpdateNextClearTime();
+}
+
+void SpatialSpanIdStore::ComponentAdd(const Worker_Op& Op)
+{
+	const Worker_AddComponentOp& AddComponentOp = Op.op.add_component;
+	EntityComponentId Id(AddComponentOp.entity_id, AddComponentOp.data.component_id);
+	Schema_Object* ComponentObject = Schema_GetComponentDataFields(AddComponentOp.data.schema_type);
+
+	TArray<uint32> UpdatedFieldIds;
+	UpdatedFieldIds.SetNumUninitialized(Schema_GetUniqueFieldIdCount(ComponentObject));
+	Schema_GetUniqueFieldIds(ComponentObject, UpdatedFieldIds.GetData());
+
+	for (uint32 FieldId : UpdatedFieldIds)
+	{
+		WriteSpanId(Id, FieldId, Op.span_id);
+	}
+}
+
+bool SpatialSpanIdStore::ComponentRemove(const Worker_Op& Op)
+{
+	const Worker_RemoveComponentOp& RemoveComponentOp = Op.op.remove_component;
+	EntityComponentId Id(RemoveComponentOp.entity_id, RemoveComponentOp.component_id);
+	return DropSpanIds(Id);
+}
+
+TArray<SpatialSpanIdStore::FieldSpanIdUpdate> SpatialSpanIdStore::ComponentUpdate(const Worker_Op& Op)
+{
+	const Worker_ComponentUpdateOp& ComponentUpdateOp = Op.op.component_update;
+	EntityComponentId Id(ComponentUpdateOp.entity_id, ComponentUpdateOp.update.component_id);
+	Schema_Object* ComponentObject = Schema_GetComponentUpdateFields(ComponentUpdateOp.update.schema_type);
+
+	TArray<uint32> UpdatedFieldIds;
+	UpdatedFieldIds.SetNumUninitialized(Schema_GetUniqueFieldIdCount(ComponentObject));
+	Schema_GetUniqueFieldIds(ComponentObject, UpdatedFieldIds.GetData());
+
+	TArray<FieldSpanIdUpdate> FieldCollisions;
+	for (uint32 FieldId : UpdatedFieldIds)
+	{
+		FieldIdMap* SpanIdMap = EntityComponentFieldSpanIds.Find(Id);
+		if (SpanIdMap != nullptr)
+		{
+			EntityComponentFieldIdUpdateSpanId* SpawnIdRef = SpanIdMap->Find(FieldId);
+			if (SpawnIdRef != nullptr)
+			{
+				FieldSpanIdUpdate Update;
+				Update.FieldId = FieldId;
+				Update.NewSpanId = Op.span_id;
+				Update.OldSpanId = SpawnIdRef->SpanId;
+				FieldCollisions.Add(Update);
+			}
+		}
+
+		WriteSpanId(Id, FieldId, Op.span_id);
+	}
+
+	return FieldCollisions;
+}
+
+void SpatialSpanIdStore::WriteSpanId(const EntityComponentId& Id, const uint32 FieldId, Trace_SpanId SpanId)
+{
+	FieldIdMap& SpanIdMap = EntityComponentFieldSpanIds.FindOrAdd(Id);
+	EntityComponentFieldIdUpdateSpanId& SpawnIdRef = SpanIdMap.FindOrAdd(FieldId);
+	SpawnIdRef.SpanId = SpanId;
+	SpawnIdRef.UpdateTime = FDateTime::Now();
```

Comment:
This can be quite a heavy call on Linux, would an incrementing counter work instead?

Reply:
Ah, see there is a seconds timeout below

Reply:
This will can be removed if we want to implement another little feature in event tracing. I would still imagine you would need to time spanids and get rid of them at some point, just on a smaller scale.